src/main/java/xyz/nucleoid/dungeons/dungeons/util/item/DgItemUtil.java

```diff
@@ -131,62 +144,112 @@ private static void finishWeaponTooltip(ItemStack stack) {
+                                "Ranged damage modifier",
+                                ((DgRangedWeapon) item).getRangedDamage(stack), EntityAttributeModifier.Operation.ADDITION
+                        ), slot
+                );
+                stack.addAttributeModifier(
+                        DgEntityAttributes.GENERIC_DRAW_TIME,
+                        new EntityAttributeModifier(
+                                DRAW_TIME_MODIFIER_ID, "Draw time modifier",
+                                ((DgRangedWeapon) item).getDrawTime(stack),
+                                EntityAttributeModifier.Operation.ADDITION
+                        ), slot
+                );
             }
+        }
+
+        if (item instanceof DgAttributeProvider) {
+            Set<EquipmentSlot> slots = ((DgAttributeProvider) item).getValidSlots(stack);
+            for (EquipmentSlot slot : slots) {
+                Multimap<EntityAttribute, EntityAttributeModifier> multimap = stack.getAttributeModifiers(slot);
+                if (!multimap.isEmpty()) {
+                    addLore(stack, LiteralText.EMPTY);
+                    if (slots.size() > 1) {
+                        addLore(stack, new TranslatableText("item.modifiers." + slot.getName()).formatted(Formatting.GRAY));
+                    }
 
-            DecimalFormatSymbols otherSymbols = new DecimalFormatSymbols(Locale.ENGLISH);
-            otherSymbols.setDecimalSeparator('.');
-            DecimalFormat format = new DecimalFormat("0.#", otherSymbols);
+                    for (Map.Entry<EntityAttribute, EntityAttributeModifier> entry : multimap.entries()) {
+                        EntityAttributeModifier modifier = entry.getValue();
+                        double value = modifier.getValue();
+                        String translationKey = entry.getKey().getTranslationKey();
+                        if (modifier.getId().equals(MELEE_DAMAGE_MODIFIER_ID)) {
+                            value += 0.5;
+                            value += EnchantmentHelper.getAttackDamage(stack, EntityGroup.DEFAULT);
+                            translationKey = DgEntityAttributes.translationKeyOf("generic.melee_damage");
+                        } else if (modifier.getId().equals(SWING_SPEED_MODIFIER_ID)) {
+                            value += 4.0;
+                            translationKey = DgEntityAttributes.translationKeyOf("generic.swing_speed");
+                        }
 
-            for (DgWeaponStat stat : stats) {
-                double value = stat.getValue();
-                addLore(stack, new TranslatableText(stat.getTranslationKey(), (value >= 0 ? "+" : "") + format.format(value)).styled(style -> style.withItalic(false).withColor(Formatting.BLUE)));
+                        double g;
+                        if (modifier.getOperation() != EntityAttributeModifier.Operation.MULTIPLY_BASE && modifier.getOperation() != EntityAttributeModifier.Operation.MULTIPLY_TOTAL) {
```

Comment:
What is this doing?

Reply:
This is just a modifier reimplementation of vanilla's attribute tooltip code